### ISSUE TITLE
VM service provides isolates with "resume" pauseEvents when started with pause_isolates_on_start

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#42875]
When using the latest nightly build, isolates show up with a pauseEvent of `resume` when starting paused. This prevents the VS Code debugger from initialising correctly as it's waiting for paused isolates in order to initialise.

```
Spawning /Users/danny/Dev/Dart SDKs/nightly-2020-07-29/bin/dart with args ["--enable-vm-service=0","--pause_isolates_on_start=true","--write-service-info=file:///var/folders/b1/ftm2whhj4mb3_c4nfjhq23z00000gn/T/dart-vm-service-f3c8.json","-DSILENT_OBSERVATORY=true","--enable-asserts","tool/lsp_spec/generate_all.dart"]
..  in /Users/danny/Dev/Google/dart-sdk/sdk/pkg/analysis_server
    PID: 21023
Starting to poll for file /var/folders/b1/ftm2whhj4mb3_c4nfjhq23z00000gn/T/dart-vm-service-f3c8.json
Successfully read JSON from /var/folders/b1/ftm2whhj4mb3_c4nfjhq23z00000gn/T/dart-vm-service-f3c8.json
Stopping polling for file /var/folders/b1/ftm2whhj4mb3_c4nfjhq23z00000gn/T/dart-vm-service-f3c8.json
Initialising debugger for ws://127.0.0.1:55958/791MZqMXV78=/ws
Connecting to VM Service at ws://127.0.0.1:55958/791MZqMXV78=/ws
==> {"id":"0","jsonrpc":"2.0","method":"getVersion","params":{}}
<== {"jsonrpc":"2.0", "result":{"type":"Version","major":3,"minor":36,"_privateMajor":0,"_privateMinor":0},"id":"0"}
==> {"id":"1","jsonrpc":"2.0","method":"getVM","params":{}}
<== {"jsonrpc":"2.0", "result":{"type":"VM","name":"vm","architectureBits":64,"hostCPU": // snip
==> {"id":"2","jsonrpc":"2.0","method":"streamListen","params":{"streamId":"Isolate"}}
==> {"id":"3","jsonrpc":"2.0","method":"streamListen","params":{"streamId":"Extension"}}
==> {"id":"4","jsonrpc":"2.0","method":"streamListen","params":{"streamId":"Debug"}}
==> {"id":"5","jsonrpc":"2.0","method":"streamListen","params":{"streamId":"Service"}}
<== {"jsonrpc":"2.0","id":"2","result":{"type":"Success"}}
<== {"jsonrpc":"2.0","id":"3","result":{"type":"Success"}}
<== {"jsonrpc":"2.0","id":"4","result":{"type":"Success"}}
<== {"jsonrpc":"2.0","id":"5","result":{"type":"Success"}}
==> {"id":"6","jsonrpc":"2.0","method":"streamListen","params":{"streamId":"Logging"}}
<== {"jsonrpc":"2.0","id":"6","result":{"type":"Success"}}
==> {"id":"7","jsonrpc":"2.0","method":"getIsolate","params":{"isolateId":"isolates/2595005099888867"}}
<== {
    "jsonrpc": "2.0",
    "result": {
        "type": "Isolate",
        "id": "isolates\/2595005099888867",
        "name": "dartdev",
        "number": "2595005099888867",
        "_originNumber": "2595005099888867",
        "startTime": 1596032780180,
        "_heaps": {/*SNIP*/},
        "runnable": true,
        "livePorts": 3,
        "pauseOnExit": false,
        "_isReloading": false,
        "pauseEvent": {
            "type": "Event",
            "kind": "Resume", // <-- ############################################
            "isolate": {
                "type": "@Isolate",
                "id": "isolates\/2595005099888867",
                "name": "dartdev",
                "number": "2595005099888867"
            },
            "timestamp": 1596032780195
        },
        // SNIP
```

cc @bkonyi - any ideas what might be wrong?

**Comment by bkonyi** [comment]
Ah this is the dartdev isolate which is similar to the kernel and service isolates such that it shouldn't be able to be paused on start. We should probably hide this isolate, especially since it's short lived, but I'm not sure why we're seeing a `Resume` event. I can't remember how I set this up off the top of my head, but I'm thinking maybe I just resumed the isolate rather than preventing it from being paused in the first place... I'll get back to you on this.

**Comment by DanTup** [comment]
I could be mistaken, but I think this was the only isolate that showed up - but the app was still paused, so it just sat there doing nothing (paused, but without VS Code seeing any paused isolates).

**Comment by bkonyi** [comment]
Are you sure there aren't additional isolates? If you're setup to build the SDK yourself, you can build a debug VM and pass in `--trace-isolates` to get an idea of isolate activity.

**Comment by DanTup** [comment]
Not certain, and not set up to build locally (or at least, haven't tried on this machine - I tend to just grab the nightlies). I'll re-test though and get a full log and post back shortly.

**Comment by DanTup** [comment]
Oh, while I remember - might not be related, but I'm intermittently seeing a failure to connect to the VM service, with the npm ws package reporting it returned a 302 (redirect) - though it doesn't give me the details. Would we ever expect a redirect connecting to it? (by default, the ws library does not follow redirects).

```
Connecting to VM Service at ws://127.0.0.1:59127/nYVaZHMZukI=/ws
Unable to start debugging: Error: Unexpected server response: 302
Exited (sigint)
```

**Comment by bkonyi** [comment]
> Oh, while I remember - might not be related, but I'm intermittently seeing a failure to connect to the VM service, with the npm ws package reporting it returned a 302 (redirect) - though it doesn't give me the details. Would we ever expect a redirect connecting to it? (by default, the ws library does not follow redirects).
> 
> ```
> Connecting to VM Service at ws://127.0.0.1:59127/nYVaZHMZukI=/ws
> Unable to start debugging: Error: Unexpected server response: 302
> Exited (sigint)
> ```

Yes, the VM service will redirect web socket connections to DDS, so that's expected. It's also a bit of an issue since the `dart:html` `WebSocket` doesn't seem to allow redirects so tooling has to do a weird hack for the forwarding to work.

**Comment by DanTup** [comment]
> so tooling has to do a weird hack for the forwarding to work

Ah, ok - what are other tools doing? This will be breaking for VS Code users until a fix is out. Though I'm not sure why it's only happening intermittently? 🤔

Here's a full log - as far as I can tell only the dartdev isolate is showing up:

[full_log.txt](https://github.com/dart-lang/sdk/files/4996204/full_log.txt)

**Comment by bkonyi** [comment]
> > so tooling has to do a weird hack for the forwarding to work
> 
> Ah, ok - what are other tools doing? This will be breaking for VS Code users until a fix is out (I can probably just enabled `followRedirects`). Though I'm not sure why it's only happening intermittently? 🤔

Observatory is currently passing a custom protocol `implicit-redirect` which manually forwards the connection to DDS. It's terribly inefficient, so I'm hoping to add SSE support for web clients instead (assuming it allows for redirects). Devtools hasn't been updated yet to use either solution.
> 
> Here's a full log - as far as I can tell only the dartdev isolate is showing up:
> 
> [full_log.txt](https://github.com/dart-lang/sdk/files/4996204/full_log.txt)

Hm, yeah I don't see any other isolates. Have you tried connecting to Observatory to inspect what's actually running?

**Comment by DanTup** [comment]
![Screenshot 2020-07-29 at 17 47 23](https://user-images.githubusercontent.com/1078012/88828765-995c1c80-d1c3-11ea-8d66-bf250c9b805a.png)

Hmmm, this looks normal - a single paused isolate (I don't even see the dartdev one) 🤔

**Comment by DanTup** [comment]
I haven't seen the 302 failure since enabling `followRedirects` (still not sure why it was intermittent), though it didn't otherwise change any behaviour.

**Comment by bkonyi** [comment]
You might just be requesting the isolates too early. Have you tried a follow-up request to check the live isolates? The dartdev isolate is extremely short lived and is killed before the main isolate comes up.

**Comment by DanTup** [comment]
Hmm, good question. We subscribe to the stream to get them - however it looks like we get the ones that are live first, then subscribe - but that leaves a race. We should be subscribing and then getting a snapshot of existing ones. I'll have a go at changing this and let you know if it fixes things. Thanks!

**Comment by DanTup** [comment]
Nope, it's not that. Here's a log where we subscribe to the debug/isolates streams before we send the getVM() request, but we still only get the dartdev isolate in that response, and we don't get events for any other isolate being created:

[full_log_2.txt](https://github.com/dart-lang/sdk/files/4996361/full_log_2.txt)

However I did notice that it's quite intermittent - sometimes it works, and we actually only see the app isolate (and it's paused). When it works, the isolate we see in the getVM response is already the app isolate (and it's paused). When it fails, the isolate we see is the dartdev one and it's "Resume".

**Comment by bkonyi** [comment]
This might be a DDS related failure then. If you can build the SDK, try and remove [these lines](https://github.com/dart-lang/sdk/blob/master/pkg/dartdev/lib/src/commands/run.dart#L122) related to the `DartDevelopmentService`. There might be a race between when the VM service starts advertising and DDS connects.

If you can't build right now, can you try passing `--disable-dart-dev` and see if that resolves the issue? I'm guessing it will, but that will at least narrow down the issue. We might have to disable DDS, so it would be great to figure out if it's causing this issue so we can disable before we branch for beta next week.

**Comment by DanTup** [comment]
Yeah, passing `--disable-dart-dev` makes it work fine every time.

If you want to try to repro locally for easier debugging, I just used the stagehand simple-console template, opened the project in VS Code, added a breakpoint to the `print` line and pressed `F5` (using a nightly build). To get logs you can run **Dart: Capture Logs** (probably untick **Analyzer** as you'll get a lot of unrelated spam).

When it works, the breakpoint will be hit. When it doesn't, it'll appear to just hang after printing "Connected to VM Service at ..." in the debug console.

**Comment by bkonyi** [comment]
Ah okay, thanks for the instructions! I always forget how to get debug information from the VSCode plugin :-) I'll take a look today and report my findings, but probably won't have a fix yet.

**Comment by bkonyi** [comment]
Okay, so disabling DDS seems to resolve the issue. None of our tools are using DDS yet, so I think I'll just disable this as to not block the beta release while I'm working on a proper fix.

**Comment by DanTup** [comment]
I was thinking more about this... Is Dart spawning a separate process (not isolate), and then forwarding the VM service to that?

If so, is the issue that we're picking up the Observatory URI from the main process (either from being printed to stdout, or using `--write-servce-info` and therefore we're not seeing the (paused) isolate for the app at all?

Could the Observatory URI be suppressed in the initial process (and also, don't write to the file in `--write-service-info`), and then printed/written only by the main process? That way we'd never connect to the VM service for the first process, and there would be no need for redirects etc.?

**Comment by bkonyi** [comment]
> I was thinking more about this... Is Dart spawning a separate process (not isolate), and then forwarding the VM service to that?
> 
> If so, is the issue that we're picking up the Observatory URI from the main process (either from being printed to stdout, or using `--write-servce-info`) and therefore we're not seeing the (paused) isolate for the app at all (leaving it paused, which is why we seem to hang)?
> 
> Could the Observatory URI be suppressed in the initial process (and also, don't write to the file in `--write-service-info`), and printed/written only by the main (second) process? That way we would never connect to the VM service for the first process, and there would be no need for redirects etc.?

Yes, it is. I investigated suppressing the Observatory URI in favor of the DDS URI but it causes quite a headache and has a number of unintended consequences. A major issue that came up is the situation where a user wants to attach to a running Flutter application. To do that, the user needs to have the URI of the VM service running on the device since DDS will have shutdown previously.

Redirecting is probably the most transparent way to handle this issue, but apparently that's an issue for the `WebSocket` implementation in most browsers. I'm working at getting SSE support working via the VM service which might be a better solution anyway as the Google internal network has some weird issues with websocket connections.

**Comment by DanTup** [comment]
> To do that, the user needs to have the URI of the VM service running on the device since DDS will have shutdown previously.

If DDS shuts down but is redirecting, wouldn't that prevent new connections to the VM service? (for example if the user wants to open DevTools)?

FWIW, Dart-Code can't currently use SSE (and the performance has been worse than WebSockets for the Dart Debug extension) so WebSockets will still need to be supported (though the node ws package appears to support redirects).

Btw - will `--write-service-info` continue to work - will just the first process write its into to the supplied file, or will it get stamped over by the second one?

**Comment by DanTup** [comment]
> The redirect only happens when DDS is connected. When it disconnects, the VM service takes control again.

I'm not very familiar with how this works so I don't completely understand this :-) Are there any notes about how this works that I can read to avoid asking lots of questions? It'd be good for me to understand how this all hangs together to help with debugging if I see any issues.

> I was unaware there were performance implications for SSE when compared to WebSockets

SSE results in a lot of HTTP round trips, so it's most noticeable when there's a lot of latency. There's some info in https://github.com/dart-lang/webdev/issues/975 and https://github.com/dart-lang/webdev/issues/993 that might be of interest. Switching to WebSockets for the Debug extension made a significant difference on GitPod for ex.

**Comment by bkonyi** [comment]
We have a [section in the service protocol](https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md#middleware-support) that covers some of the details. There's also a fairly bare-bones [DDS protocol document](https://github.com/dart-lang/sdk/blob/master/pkg/dds/dds_protocol.md) which describes the service extensions offered by DDS. I'm happy to answer any questions, so feel free to ask them or setup a GVC where we can discuss (just not too early in the morning please :-)).

Ah, I didn't even think about that. If that's the case, a couple of extra requests to figure out the DDS URI seems fairly reasonable in comparison.

**Comment by DanTup** [comment]
Aha, that helps - especially the note under [single client mode](https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md#single-client-mode).

So if I understand correctly, when we run with dartdev enabled, what's happening is:

- We run `dart` and pass it a script to run
- It instead runs dartdev(?) code, which spawns another VM to run the script
- The dartdev/DDS process connects to the VM service for the second process and puts the VM service into single user mode, which causes all connections to be routed back to DDS (which t then proxies on to the original VM service in the second process, or handles itself if they're calls to its extensions)

If so, I have some additional questions:

- Why DDS instead of extending the VM service directly (or with service extensions)?
- How can VS Code end up connected to a VM service that shows the dartdev isolates? It seems like it should either connect to the second service (bypassing DDS) or be redirecting to DDS which proxies on to the second service

Actually, re-reading above, you said:

> The dartdev isolate is extremely short lived and is killed before the main isolate comes up

But I don't understand how that fits in with the above (since this talked about a short-lived isolate, but my understanding above is a long-running process?).

**Comment by bkonyi** [comment]
> Aha, that helps - especially the note under [single client mode](https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md#single-client-mode).
> 
> So if I understand correctly, when we run with dartdev enabled, what's happening is:
> 
> * We run `dart` and pass it a script to run
> * It instead runs dartdev(?) code, which spawns another VM to run the script
> * The dartdev/DDS process connects to the VM service for the second process and puts the VM service into single user mode, which causes all connections to be routed back to DDS (which t then proxies on to the original VM service in the second process, or handles itself if they're calls to its extensions)

That's how this used to work, but we've instead moved to running DDS in a separate process which connects to the service in the initial process that's running the provided script. Once DDS connects, the VM service enters single client mode and will forward any web socket connections to DDS.

> If so, I have some additional questions:
> 
> * Why DDS instead of extending the VM service directly (or with service extensions)?

DDS is meant to be middleware that can run in a separate process from the VM service. There's a few benefits to this:

* We've been really hesitant to add functionality to the VM service that really doesn't have any reason to be tightly integrated with the VM. Features like client synchronization don't actually require any knowledge about VM state that isn't already available over the service protocol (e.g., isolate state). 
* We've also been weary of adding functionality to the service protocol that has the potential to have a high impact on resources consumed by the process. For example, while introducing a log history buffer so clients can receive historical logs from before they had connected would be extremely useful, buffering logs in process would result in memory consumption appearing much larger than it actually would be if log history wasn't tracked. Running out of process means that we can provide functionality without skewing performance numbers of the inspected process.
* Many feature requests are for features that would also be useful when when targeting other VM service implementations like dwds. Implementing these features outside of the VM makes them easier to share with dwds.
* The VM service code is tightly integrated with the VM and isn't the easiest code to work on if you're not already familiar with the workings of the VM. Writing DDS as a standalone Dart program means we can take advantage of packages in the Dart ecosystem and also encourage other non-VM team members to contribute features that they need.

> * How can VS Code end up connected to a VM service that shows the dartdev isolates? It seems like it should either connect to the second service (bypassing DDS) or be redirecting to DDS which proxies on to the second service
> 
> Actually, re-reading above, you said:
> 
> > The dartdev isolate is extremely short lived and is killed before the main isolate comes up
> 
> But I don't understand how that fits in with the above (since this talked about a short-lived isolate, but my understanding above is a long-running process?).

Dartdev runs in the initial process before running user code and should be killed / shutting down by the time the main isolate is started.

**Comment by DanTup** [comment]
@bkonyi thanks, everything is much clearer now :-)

Back to the original issue, I'm still not sure why if we connect too early (when the dartdev isolate is alive) things don't work. But would it be possible to just delay printing the VM service URI (and writing the `--write-service-info` file) until after `dartdev` is done and the new isolate is alive to avoid the possibility?

**Comment by bkonyi** [comment]
> @bkonyi thanks, everything is much clearer now :-)

Glad that helped! :-)

> Back to the original issue, I'm still not sure why if we connect too early (when the dartdev isolate is alive) things don't work. But would it be possible to just delay printing the VM service URI (and writing the `--write-service-info` file) until after `dartdev` is done and the new isolate is alive to avoid the possibility?

Yes, that's the plan, I just haven't gotten around to it yet. See https://github.com/dart-lang/sdk/issues/42727 to track that change.

**Comment by DanTup** [comment]
> Yes, that's the plan, I just haven't gotten around to it yet. See #42727 to track that change.

Super, thanks!

**Comment by mit-mit** [comment]
@bkonyi sounds like either this bug or @42727 should be tagged for Sept. release?

**Comment by DanTup** [comment]
@bkonyi I'm not sure if it's related to the same issue, but Dart-Code's integration tests are failing on latest dev SDK (and I can repro locally using the latest nightly) with lots of this error in response to `setLibraryDebuggable` calls:

```
could not send message [x] to isolate
```

Seems like the first two worked, but the rest failed. The isolate's name is `dartdev` so I suspect it might be related to this/#42727 and it's all going to the wrong isolate. If not and it's better to raise a separate issue, let me know!

[full log](https://github.com/dart-lang/sdk/files/5051882/dart_cli_debugger_receives_the_expected_output.txt)

**Comment by bkonyi** [comment]
I'm guessing the dartdev isolate had exited sometime after the first couple of `setLibraryDebuggable` calls. You're probably going to want to filter the dartdev isolate and not perform any operations on it for now.

I plan on adding a new property to isolates which can be used to designate them as "system" isolates. You'll want to filter your isolate operations on this property once that change lands since we're also going to be making some other system isolates visible which are currently hidden (e.g., kernel-service and vm-service).

**Comment by DanTup** [comment]
> I plan on adding a new property to isolates which can be used to designate them as "system" isolates

If you let me know when this is available, I can add filters. I've disabled the Dart-Code tests on dev SDK builds for now as they were very flaky (and some very slow due to things timing out). I'll re-enable them once this is done. Thanks!

**Comment by devoncarew** [comment]
Is this only an issue when running with DDS? If so, we can move this out of the Sept. milestone (and into the next milestone I assume).

**Comment by a-siva** [comment]
There is a CL in progress that provides a new property ("system" isolates) that allows for tools to filter out isolate operations based on this property https://dart-review.googlesource.com/c/sdk/+/158063

**Comment by bkonyi** [comment]
@DanTup this should be resolved now that [this CL](https://dart-review.googlesource.com/c/sdk/+/158063) has landed. The dartdev isolate has been marked as a system isolate and won't appear in the `isolates` list in the `VM` object, but will instead appear in a new `systemIsolates` list along with the service and kernel service isolates. You can check if an isolate is a system isolate by checking the `isSystemIsolate` property on any `@Isolate` or `Isolate`.

**Comment by DanTup** [comment]
@bkonyi great, thanks! I can't currently reproduce the issue here when using the latest nightly build (though I'm not sure whether DDS is still disabled, or whether that matters).

**Comment by bkonyi** [comment]
Nope, DDS is still disabled but it shouldn't matter since DartDev is still spawned in an isolate regardless of whether or not DDS is enabled. Glad to hear it's fixed!

**Comment by DanTup** [comment]
Great! I'll re-enable Dart-Code's integration tests using the dev channel once there's a dev SDK build with this in and let you know if I still see failures there. Thanks!
